pkg/terra/contract_reader.go

```diff
@@ -0,0 +1,261 @@
+package terra
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+	"time"
+
+	cosmosSDK "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/smartcontractkit/libocr/offchainreporting2/types"
+
+	"github.com/smartcontractkit/chainlink-terra/pkg/terra/client"
+)
+
+type OCR2Reader struct {
+	address     cosmosSDK.AccAddress
+	chainReader client.Reader
+	lggr        Logger
+}
+
+func NewOCR2Reader(addess cosmosSDK.AccAddress, chainReader client.Reader, lggr Logger) *OCR2Reader {
+	return &OCR2Reader{
+		address:     addess,
+		chainReader: chainReader,
+		lggr:        lggr,
+	}
+}
+
+func (or *OCR2Reader) fetchLatestConfigDetails(ctx context.Context) (changedInBlock uint64, configDigest types.ConfigDigest, err error) {
```

Comment:
I'd suggest we make these functions public and drop the `fetch*` prefix. The reader could be useful for other components of the system (monitoring, test assertions, etc.).

```suggestion
func (r *OCR2Reader) LatestConfigDetails(ctx context.Context) (changedInBlock uint64, configDigest types.ConfigDigest, err error) {
```

Reply:
I did the same with `ContractCache`, and embedded them as well to eliminate the boilerplate method forwarding (and the whole of `MedianContract` - the cache implements the interface directly now). Much simpler overall, but happy to switch back to explicit forwarding if embedding is confusing or error prone.